# no-missing-link-fragments

Disallow link fragments that don't exist in the document.

## Rule Details

This rule verifies that all link fragments (using the `#` syntax in URLs) reference valid headings that exist in the document. Link fragments are commonly used to navigate to specific sections within a document, and when they're invalid, users won't be directed to the intended location.

### Options

This rule has an options object with the following properties (parameter naming follows markdownlint's MD051 rule for compatibility):

* `ignoreCase`: Boolean, default `false` - When `true`, fragment case will be ignored when comparing to heading IDs
* `allowPattern`: String, default `""` - Regular expression pattern for fragments to ignore

Example configuration:

```json
{
  "markdown/no-missing-link-fragments": ["error", {
    "ignoreCase": true,
    "allowPattern": "^section-"
  }]
}
```

### Examples

#### ❌ Incorrect

```md
[Invalid link](#non-existent-section)

## Introduction

[Also invalid](#INTRODUCTION)
```

In the examples above:
- `#non-existent-section` does not match any heading in the document
- `#INTRODUCTION` doesn't match the actual heading ID which would be `#introduction` (unless `ignoreCase` is set to `true`)

#### ✅ Correct

```md
[Valid link](#introduction)

## Introduction

[Also valid](#heading-with-spaces)

## Heading with spaces

<!-- With ignoreCase: true -->
[Case insensitive link](#INTRODUCTION)

## Introduction

<!-- With allowPattern: "^section-" -->
[Ignored pattern link](#section-123)
```

### Further Reading

This rule is similar to the markdownlint MD051 rule, which checks that link fragments are valid. It ensures that all links that use fragment identifiers point to actual headings in the document.

## When Not To Use It

You might want to disable this rule if you're linking to fragments that are generated by other means, such as manually created anchor elements in HTML within the markdown. Alternatively, you can use the `allowPattern` option to exclude specific fragment patterns from validation. 